### PR TITLE
[libcdio]Add new port

### PR DESCRIPTION
**Pull request by bwrsandman**
Implements  #2347

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by bwrsandman** [comment]
Note to maintainers: This library has libtools dependencies like alsa does. I'm not sure how to manage these dependencies.

File: ports/libcdio/0001-Remove-doc-from-config-and-make.patch [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+diff --git a/Makefile.am b/Makefile.am
+index 24769c60..76de6dfb 100644
+--- a/Makefile.am
++++ b/Makefile.am
+@@ -40,7 +40,7 @@ EXTRA_DIST = \
+ 	make-check-filter.rb \
+ 	package/libcdio.spec.in
+ 
+-SUBDIRS = doc include lib src test example
++SUBDIRS = include lib src test example
```
Comment (dg0yt) [inline ports/libcdio/0001-Remove-doc-from-config-and-make.patch:21]:
```suggestion
+SUBDIRS =     include lib src
```
if it doesn't complain.

File: ports/libcdio/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    0003-Fix-free-while-still-in-use-in-iso9660.hpp.patch
+    0004-src-cdda-player.c-always-use-s-style-format-for-prin.patch
+)
+
+vcpkg_configure_make(
+    SOURCE_PATH "${SOURCE_PATH}"
+    AUTOCONFIG
+    OPTIONS
+        ${OPTIONS}
+        --prefix=${CURRENT_INSTALLED_DIR}
```
Comment (dg0yt) [inline ports/libcdio/portfile.cmake:22]:
```suggestion
```

File: ports/libcdio/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        --disable-example-progs
+        --disable-dependency-tracking
+        --enable-maintainer-mode
```
Comment (dg0yt) [inline ports/libcdio/portfile.cmake:25]: Probably not.
```suggestion
        --enable-maintainer-mode
```

File: ports/libcdio/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+vcpkg_install_make()
+
+vcpkg_copy_pdbs()
+
+vcpkg_fixup_pkgconfig()
+vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/cdio/cdio_config.h" "${CURRENT_BUILDTREES_DIR}" "")
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+
+configure_file("${CMAKE_CURRENT_LIST_DIR}/vcpkg-cmake-wrapper.cmake" "${CURRENT_PACKAGES_DIR}/share/${PORT}/vcpkg-cmake-wrapper.cmake" @ONLY)
```
Comment (dg0yt) [inline ports/libcdio/portfile.cmake:38]: You may implement an `unofficial-libcdio-config.cmake` to provide an `unofficial::libcdio::libcdio`. (The prefixes matter.)
But please don't abuse the wrapper feature if there is nothing to wrap.

And maybe the official pkgconfig is enough.

**Review by dg0yt — CHANGES_REQUESTED** [review]
Some early comments.

**Comment by bwrsandman** [comment]
I've got this somewhat compiling on windows using the same autoconf path as mac and linux...

However, I think I have run into a libtool bug...

```
libtool: link:  compile cl.exe -o .libs\\iso9660-11.dll  .libs/iso9660.obj .libs/iso9660_fs.obj .libs/rock.obj .libs/xa.obj   -nologo -W3 -utf-8 -MP -MDd -Z7 -Ob0 -Od -RTC1 -Xlinker -LC:/path/to/vcpkg/installed/x64-windows/debug/lib -Xlinker -machine:x64 -Xlinker -nologo -Xlinker -debug -Xlinker -INCREMENTAL    ../../lib/driver/.libs/cdio.lib -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -lcomdlg32 -ladvapi32 -lwinmm "@.libs\\iso9660-11.dll.exp" -Wl,-DLL,-IMPLIB:".libs\\iso9660.lib"
LINK : warning LNK4044: unrecognized option '/LC:/path/to/vcpkg/installed/x64-windows/debug/lib'; ignored
   Creating library .libs\iso9660.lib and object .libs\iso9660.exp
iso9660_fs.obj : error LNK2001: unresolved external symbol CDIO_SECTOR_SYNC_HEADER
```

What's happening here is that `-L` is not being substituted for `/LIBPATH:`.

Debugging the compile command (added echo as a cl.exe command)
```
$ ./compile echo -LC:/path/to/vcpkg/installed/x64-windows/debug/lib
-link -LIBPATH:C:/path/to/vcpkg/installed/x64-windows/debug/lib
$ ./compile echo -L C:/path/to/vcpkg/installed/x64-windows/debug/lib
-link -LIBPATH:C:/path/to/vcpkg/installed/x64-windows/debug/lib
$ ./compile echo -Xlinker -LC:/path/to/vcpkg/installed/x64-windows/debug/lib
-link -LC:/path/to/vcpkg/installed/x64-windows/debug/lib
$ ./compile echo -Xlinker -L C:/path/to/vcpkg/installed/x64-windows/debug/lib
C:/path/to/vcpkg/installed/x64-windows/debug/lib -link -L
$ ./compile echo -I. -LC:/path/to/vcpkg/installed/x64-windows/debug/lib
-I. -link -LIBPATH:C:/path/to/vcpkg/installed/x64-windows/debug/lib
$ ./compile echo -I. -L C:/path/to/vcpkg/installed/x64-windows/debug/lib
-I. -link -LIBPATH:C:/path/to/vcpkg/installed/x64-windows/debug/lib
```

Clearly this shows that `-Xlinker` prevents the proper substitution of `-L` to `/LIBPATH:`.

I looked pretty hard in the libcdio code for where `-Xlinker` gets added  or where `compile` comes from but have come up short so far.

If anyone has pointers on what to do, it would help me a great deal.

**Comment by dg0yt** [comment]
The `-Xlinker` may come from  `vcpkg_configure_make` which is tuned for standard autotools projects with libtool and for using the `compile` wrapper which comes with vcpkg. And which has a quirk concerning this flag (#31228 C4).
I'm surprised to see a gnulib unistd header in the port. Isn't it included in upstream's distribution?

**Comment by bwrsandman** [comment]
> The -Xlinker may come from vcpkg_configure_make which is tuned for standard autotools projects with libtool and for using the compile wrapper which comes with vcpkg. And which has a quirk concerning this flag (https://github.com/microsoft/vcpkg/pull/31228 C4).

Ok so that would be where it's added...

The issue about `-Xlinker -L` also exists outside of ` vcpkg_configure_make`.

The `compile` script that consumes -Xlinker skips treatment of following flags:
```sh
	-L*)
	  func_cl_dashL "${1#-L}"
	  ;;
	-static)
	  shared=false
	  ;;
	-Xlinker)
	  eat=1
	  linker_opts="$linker_opts $2" # skips func_cl_dashL of the next argument if it starts with -L
	  ;;
```

> I'm surprised to see a gnulib unistd header in the port. Isn't it included in upstream's distribution?

I was surprised not to find it in upstream, because they mention a `missing/` directory with `stdint.h`, `inttypes.h` and
`unistd.h` in `MSVC/README`. However, this same readme mentions msvc 2010, so it's quite old. I also looked throughout the releases for a history of these files and could not find one.

**Comment by dg0yt** [comment]
`gnulib` is not meant to be used as a separate lib, but provides scripts for upstream maintainers to copy the relevant modules into upstream packages.

**Comment by JonLiu1993** [comment]
@bwrsandman, Is work still being done for this PR?

**Comment by bwrsandman** [comment]
I would still like to get this merged but I haven't had time to work on it at the moment.

**Comment by JonLiu1993** [comment]
@bwrsandman, Since there has been no progress on this PR for nearly a year, I will temporarily close this PR. If you have updates, please feel free to let me know and open it again..

**Pull request closed without merging**